### ISSUE TITLE
[java] UnnecessaryCast false-positive for raw types

### ISSUE THREAD

**Issue by blacelle** [pmd/pmd#4822]
**Affects PMD Version:** 
7.0.0-rc4

**Rule:** [UnnecessaryCast](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessarycast)

**Description:**
False-positive as we need to cast without diamond operator to compile.

**Code Sample demonstrating the issue:**

```java
import java.util.Objects;
import java.util.Comparator;
import static java.util.Comparator.nullsFirst;

public class Foo {
    public int compare(Comparable<?> o1, Comparable<?> o2) {
        final Comparator nullComparator = nullsFirst(null);
        if (o1 == null || o2 == null) {
            return Objects.compare(o1, o2, nullComparator);
        }
        Class<?> c1 = o1.getClass();
        Class<?> c2 = o2.getClass();
        if (c1 != c2) {
            return c1.getName().compareTo(c2.getName());
        }

        return ((Comparable) o1).compareTo(o2); // <--- Unnecessary cast (Comparable)
    }
}
```

**Expected outcome:**

PMD reports a violation at line ..., but that's wrong. That's a false positive.

**Running PMD through:** Maven

**Comment by adangel** [comment]
You can write the code using generics, than you don't need a cast:

```java
    public <T extends Comparable<? super T>> int compare2(T o1, T o2) {
        final Comparator<T> nullComparator = nullsFirst(null);
        if (o1 == null || o2 == null) {
            return Objects.compare(o1, o2, nullComparator);
        }
        Class<?> c1 = o1.getClass();
        Class<?> c2 = o2.getClass();
        if (c1 != c2) {
            return c1.getName().compareTo(c2.getName());
        }

        return o1.compareTo(o2);
    }
```

I'll keep this issue open - maybe we can detect such cases, when a cast to a raw type is needed.

**Comment by adangel** [comment]
If you really want to compare to objects of different types (apples and oranges), just don't use generics at all:

```java
    public int compare3(Comparable o1, Comparable o2) {
        final Comparator nullComparator = nullsFirst(null);
        if (o1 == null || o2 == null) {
            return Objects.compare(o1, o2, nullComparator);
        }
        Class<?> c1 = o1.getClass();
        Class<?> c2 = o2.getClass();
        if (c1 != c2) {
            return c1.getName().compareTo(c2.getName());
        }

        return o1.compareTo(o2);
    }
```
